### PR TITLE
civo cli shell plugin added

### PR DESCRIPTION
**Pull request by siddhikhapare**
## Type of change
 

- [x] Created a new plugin
- [ ] Improved an existing plugin
- [ ] Fixed a bug in an existing plugin
- [ ] Improved contributor utilities or experience

## Related Issue(s)
  

* Resolves: #
* Relates: #

### CODE REVIEW

**Comment by techcraver** [comment]
Hello! Is this for the Hashnode hackathon?

**Comment by siddhikhapare** [comment]
@techcraver  Yes. @hculea helped me.

**Comment by arunsathiya** [comment]
@siddhikhapare There's already a CIVO shell plugin in review:

- https://github.com/1Password/shell-plugins/pull/296

Could you help me understand if your CIVO shell plugin is different from the one on the other PR? I noticed that the environment variables are different in both PRs too. 🤔

**Comment by siddhikhapare** [comment]
@arunsathiya  We have used CIVO_TOKEN as environment variable names by convention from here as mentioned in this file of Civo cli https://github.com/civo/cli/blob/99470155c7f5d03b80be9ce45832d8e6d54dc5ed/config/config.go#L93

File: plugins/civo/api_key.go [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+	"github.com/1Password/shell-plugins/sdk/schema"
+	"github.com/1Password/shell-plugins/sdk/schema/credname"
+	"github.com/1Password/shell-plugins/sdk/schema/fieldname"
+)
+
+func APIKey() schema.CredentialType {
+	return schema.CredentialType{
+		Name:          credname.APIKey,
+		DocsURL:       sdk.URL("https://www.civo.com/docs/account/api-keys"), 
+		ManagementURL: sdk.URL("https://www.civo.com/api#authentication"),
```
Comment (hculea) [inline plugins/civo/api_key.go:18]: Let's ensure that this field contains a user-specific page 😄 From what I see, this still links to the documentation.

File: plugins/civo/api_key.go [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+						Uppercase: true,
+						Lowercase: true,
+						Digits:    true,
+					},
+				},
+			},
+		},
+		DefaultProvisioner: provision.EnvVars(defaultEnvVarMapping),
+		Importer: importer.TryAll(
+			importer.TryEnvVarPair(defaultEnvVarMapping),
```
Comment (hculea) [inline plugins/civo/api_key.go:36]: We could also try the env-var pair that's used by `civo apikey save`, as defined here:

https://github.com/civo/cli/blob/99470155c7f5d03b80be9ce45832d8e6d54dc5ed/cmd/apikey/apikey_save.go#L33

File: plugins/civo/api_key_test.go [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+			ExpectedCandidates: []sdk.ImportCandidate{
+				{
+					Fields: map[sdk.FieldName]string{
+						fieldname.Token: "vvUmolh6TG7VMzCkhfWvOCJfpZMzM6W5hWkjurT63FoEXAMPLE",
+					},
+				},
+			},
+		},
+	})
+}
```
Comment (hculea) [inline plugins/civo/api_key_test.go:54]: Please add the required fixtures for this test to work as well 😄

File: plugins/civo/civo.go [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+)
+
+func CivoCLI() schema.Executable {
+	return schema.Executable{
+		Name:      "Civo CLI", // TODO: Check if this is correct
+		Runs:      []string{"civo"},
+		DocsURL:   sdk.URL("https://civo.com/docs/cli"), // TODO: Replace with actual URL
+		NeedsAuth: needsauth.IfAll(
+			needsauth.NotForHelpOrVersion(),
+			needsauth.NotWithoutArgs(),
```
Comment (hculea) [inline plugins/civo/civo.go:17]: Let's also opt out of authentication for `apikey save` and when `--config` is passed.

**Review by hculea — COMMENTED** [review]
Thank you for your contribution! There are a couple left TODOs, please have a look over those as well.

**Comment by techcraver** [comment]
@siddhikhapare Awesome! Be sure to write your blog post Hashnode as well to accompany this PR - that's how your Hackathon submission is complete. :)

File: plugins/civo/civo.go (hunk 4 above)
Comment (siddhikhapare) [reply plugins/civo/civo.go:17]: @hculea Thank you for you guidance. I made changes in files.

**Comment by siddhikhapare** [comment]
@techcraver Thank you. Sure.

File: plugins/civo/api_key.go [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+			TryCivoConfigFile("~/.civo.json"),
+		)}
+}
+
+var defaultEnvVarMapping = map[string]sdk.FieldName{
+	"CIVO_TOKEN": fieldname.APIKey, 
+	"CIVO_API_KEY":fieldname.APIKey,
+	"CIVO_API_KEY_NAME":fieldname.APIKeyID,
```
Comment (hculea) [inline plugins/civo/api_key.go:49]: Let's separate these in a different map, and only use these for the importers, not also the provisioner. 😄

File: plugins/civo/api_key_test.go [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+		},
+		
+		"config file": {
+			Files: map[string]string{
+				"~/.civo.json": plugintest.LoadFixture(t, "civo.json"),
+			},
+			ExpectedCandidates: []sdk.ImportCandidate{
+				{
+					Fields: map[sdk.FieldName]string{
+						fieldname.Token: "vvUmolh6TG7VMzCkhfWvOCJfpZMzM6W5hWkjurT63FoEXAMPLE",
```
Comment (hculea) [inline plugins/civo/api_key_test.go:54]: should this be `fieldname.APIKey`?

File: plugins/civo/civo.go (hunk 4 above)
Comment (hculea) [reply plugins/civo/civo.go:17]: Looking through https://www.civo.com/docs/overview/civo-cli#commands-and-help I can see a few other commands that I don't think should require authentication, such as `civo update`. Can you please have a look over the rest as well? 😄 Would love to keep this list as precise as possible.

**Comment by hculea** [comment]
@siddhikhapare I see there is a another contribution for the same plugin: https://github.com/1Password/shell-plugins/pull/296 authored by @itsCheithanya.

To deduplicate some of the remaining work and help move this forward faster, I'd love if you two could collaborate to get this to the finish line 😄 

Regardless of which PR you will choose to proceed with, we will be counting this contribution for both of you for the hackathon. 

Please let me know if that is okay with you and whether I could help you two connect, on Discord or the 1Password Developer Workspace on Slack! Let's go 🚀

**Comment by siddhikhapare** [comment]
@hculea I am grateful to collaborate for this contribution on slack.   I tried this logic but it did not work.
```golang 
func TryCivoConfigFile(path string) sdk.Importer {
	return importer.TryFile(path, func(ctx context.Context, contents importer.FileContents, in sdk.ImportInput, out *sdk.ImportAttempt) {
		var config Config
		if err := contents.ToJSON(&config); err != nil {
			out.AddError(err)
			return
		}

		if len(config.APIKeys) == 0 || len(config.SecondAuth) == 0 {
			return
		}

		for key, value := range config.APIKeys {
			out.AddCandidate(sdk.ImportCandidate{
				NameHint: key,
				Fields: map[sdk.FieldName]string{
					fieldname.APIKey: value,
				},
			})
		}

		for key, value := range config.SecondAuth {
			out.AddCandidate(sdk.ImportCandidate{
				NameHint: key,
				Fields: map[sdk.FieldName]string{
					fieldname.APIKey:   value.APIKeys,
					fieldname.APIKeyID: value.ApiKeyID,
				},
			})
		}
	})
}

type Config struct {
	APIKeys    map[string]string      `json:"apikeys"`
	SecondAuth map[string]OtherAPIKey `json:"secondauth"`
}

type OtherAPIKey struct {
	APIKeys  string `json:"secondapikey"`
	ApiKeyID string `json:"name"`
}

```

**Pull request closed without merging**